derp/xdp/xdp.c

```diff
@@ -0,0 +1,585 @@
+	// the context of the data. cilium/ebpf uses native endian encoding for map
+	// encoding even if we use big endian types here, e.g. __be16.
+	__u16 dst_port;
+};
+struct config *unused_config __attribute__((unused)); // required by bpf2go -type
+
+struct {
+      __uint(type, BPF_MAP_TYPE_ARRAY);
+      __uint(key_size, sizeof(__u32));
+      __uint(value_size, sizeof(struct config));
+      __uint(max_entries, 1);
+} config_map SEC(".maps");
+
+struct counters_key {
+	__u8 unused;
+	__u8 af;
+	__u8 pba;
+	__u8 prog_end;
+};
+struct counters_key *unused_counters_key __attribute__((unused)); // required by bpf2go -type
+
+enum counter_key_af {
+	COUNTER_KEY_AF_UNKNOWN,
+	COUNTER_KEY_AF_IPV4,
+	COUNTER_KEY_AF_IPV6,
+	COUNTER_KEY_AF_LEN
+};
+enum counter_key_af *unused_counter_key_af __attribute__((unused)); // required by bpf2go -type
+
+enum counter_key_packets_bytes_action {
+	COUNTER_KEY_PACKETS_PASS_TOTAL,
+	COUNTER_KEY_BYTES_PASS_TOTAL,
+	COUNTER_KEY_PACKETS_ABORTED_TOTAL,
+	COUNTER_KEY_BYTES_ABORTED_TOTAL,
+	COUNTER_KEY_PACKETS_TX_TOTAL,
+	COUNTER_KEY_BYTES_TX_TOTAL,
+	COUNTER_KEY_PACKETS_DROP_TOTAL,
+	COUNTER_KEY_BYTES_DROP_TOTAL,
+	COUNTER_KEY_PACKETS_BYTES_ACTION_LEN
+};
+enum counter_key_packets_bytes_action *unused_counter_key_packets_bytes_action __attribute__((unused)); // required by bpf2go -type
+
+enum counter_key_prog_end {
+	COUNTER_KEY_END_UNSPECIFIED,
+	COUNTER_KEY_END_UNEXPECTED_FIRST_STUN_ATTR,
+	COUNTER_KEY_END_INVALID_UDP_CSUM,
+	COUNTER_KEY_END_INVALID_IP_CSUM,
+	COUNTER_KEY_END_NOT_STUN_PORT,
+	COUNTER_KEY_END_INVALID_SW_ATTR_VAL,
+	COUNTER_KEY_END_LEN
+};
+enum counter_key_prog_end *unused_counter_key_prog_end __attribute__((unused)); // required by bpf2go -type
+
+#define COUNTERS_MAP_MAX_ENTRIES ((COUNTER_KEY_AF_LEN - 1) << 16) | \
+                                 ((COUNTER_KEY_PACKETS_BYTES_ACTION_LEN - 1) << 8) | \
+                                 (COUNTER_KEY_END_LEN - 1)
+
+struct {
+      __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+      __uint(key_size, sizeof(struct counters_key));
+      __uint(value_size, sizeof(__u64));
+      __uint(max_entries, COUNTERS_MAP_MAX_ENTRIES);
+} counters_map SEC(".maps");
+
+struct stunreq {
+	__be16 type;
+	__be16 length;
+	__be32 magic;
+	__be32 txid[3];
+	// attributes follow
+};
+
+struct stunattr {
+	__be16 num;
+	__be16 length;
+};
+
+struct stunxor {
+	__u8 unused;
+	__u8 family;
+	__be16 port;
+	__be32 addr;
+};
+
+struct stunxor6 {
+	__u8 unused;
+	__u8 family;
+	__be16 port;
+	__be32 addr[4];
+};
+
+#define STUN_BINDING_REQUEST 1
+
+#define STUN_MAGIC 0x2112a442
+
+#define STUN_ATTR_SW 0x8022
+
+#define STUN_ATTR_XOR_MAPPED_ADDR 0x0020
+
+#define STUN_BINDING_RESPONSE 0x0101
+
+#define STUN_MAGIC_FOR_PORT_XOR 0x2112
+
+#define MAX_UDP_LEN_IPV4 1480
+
+#define MAX_UDP_LEN_IPV6 1460
+
+#define IP_MF 0x2000
+#define IP_OFFSET 0x1fff
+
+static __always_inline __u16 csum_fold_flip(__u32 csum) {
+	__u32 sum;
+	sum = (csum >> 16) + (csum & 0xffff);
+	sum += (sum >> 16);
```

Comment:
```suggestion
	sum = (csum >> 16) + (csum & 0xffff); // maximum value 0x1fffe
	sum += (sum >> 16); // maximum value 0xffff
```

Reply:
updated in https://github.com/tailscale/tailscale/pull/12390/commits/7addbb6e73cf893d6fedba6198d32967f7308a60